joinmarket-webui/jam
# feat(Jam): add success message

Resolves #489.

This PR adds a success message after a scheduler run completes.

Before this PR, the page reloaded greeting the user with a message ala "The scheduler requires your UTXOs to have 5 or more confirmations", which is confusing to say the least. 
After this PR, the user will see a screen indicating success with a short summary, i.e. "Successfully completed n scheduled transactions".

Please note that due to polling the scheduler data, this might miss the state change of the last scheduled item. To mitigate this situation, it is checked whether only frozen UTXOs remain in the wallet, which should always be the case on mainnet. However, in "dev mode" (where – on purpose – not all jars a swept in order to decrease the time the scheduler needs to run; an on top – the funds are sent to Jar A of the same wallet) this might still occur. 
https://github.com/joinmarket-webui/jam/blob/8a0be65a185180f0561699bbed2f1fb0a4d0ae2f/src/components/Jam.tsx#L249-L260
If you have an idea on how to upgrade this from a workaround to a fully reliable fix, please say so!

As always, any feedback is greatly appreciated. As well as any improvement of  wording/phrases/etc.! 

## How to test
Start the scheduler while mining some blocks. Wait for the scheduler run to be completed.
Since polling is done every 10 seconds, and mining in the regtest environment is also done every 10 seconds, the situation described above might apply and you won't see the success message. Either increase the time between regtest blocks by running `watch -n 60 ./docker/regtest/mine-block.sh` (blocks every 60 secs) instead of `npm run regtest:mine` (blocks every 10 secs), OR send the funds to an external test wallet (e.g. to one of the regtest makers, `http://localhost:29080`). As mentioned, this should only be an issue on `regtest`, not in production.

## :camera_flash: Screenshots

<img src="https://user-images.githubusercontent.com/3358649/214094225-e8d4f87e-5efb-4545-9965-fc01d8ea1d5e.png" width=300 /> <img src="https://user-images.githubusercontent.com/3358649/214095640-9d7a0e49-71ff-4b90-a164-ed4d91ca4587.png" width=300 />
